lib/datadog/di/probe_notifier_worker.rb

```diff
@@ -0,0 +1,242 @@
+        @status_queue = []
+        @snapshot_queue = []
+        @transport = transport
+        @lock = Mutex.new
+        @wake = Core::Semaphore.new
+        @io_in_progress = false
+        @sleep_remaining = nil
+        @wake_scheduled = false
+      end
+
+      attr_reader :settings
+
+      def start
+        return if defined?(@thread) && @thread
+        @thread = Thread.new do
+          loop do
+            # TODO If stop is requested, we stop immediately without
+            # flushing the submissions. Should we send pending submissions
+            # and then quit?
+            break if @stop_requested
+
+            sleep_remaining = @lock.synchronize do
+              if sleep_remaining && sleep_remaining > 0
+                # Recalculate how much sleep time is remaining, then sleep that long.
+                set_sleep_remaining
+              else
+                0
+              end
+            end
+
+            if sleep_remaining > 0
+              # Do not need to update @wake_scheduled here because
+              # wake-up is already scheduled for the earliest possible time.
+              wake.wait(sleep_remaining)
+              next
+            end
+
+            begin
+              more = maybe_send
+            rescue => exc
+              raise if settings.dynamic_instrumentation.propagate_all_exceptions
+
+              warn "Error in probe notifier worker: #{exc.class}: #{exc} (at #{exc.backtrace.first})"
+            end
+            @lock.synchronize do
+              @wake_scheduled = more
+            end
+            wake.wait(more ? MIN_SEND_INTERVAL : nil)
+          end
+        end
+      end
+
+      # Stops the background thread.
+      #
+      # Attempts a graceful stop with the specified timeout, then falls back
+      # to killing the thread using Thread#kill.
+      def stop(timeout = 1)
+        @stop_requested = true
+        wake.signal
+        unless thread&.join(timeout)
+          thread.kill
+        end
+        @thread = nil
+      end
+
+      # Waits for background thread to send pending notifications.
+      #
+      # This method waits for the notification queue to become empty
+      # rather than for a particular set of notifications to be sent out,
+      # therefore, it should only be called when there is no parallel
+      # activity (in another thread) that causes more notifications
+      # to be generated.
+      def flush
+        loop do
+          if @thread.nil? || !@thread.alive?
+            return
+          end
+
+          io_in_progress, queues_empty = @lock.synchronize do
+            [io_in_progress?, status_queue.empty? && snapshot_queue.empty?]
+          end
+
+          if io_in_progress
+            # If we just call Thread.pass we could be in a busy loop -
+            # add a sleep.
+            sleep 0.25
+            next
+          elsif queues_empty
+            break
+          else
+            sleep 0.25
+            next
```

Comment:
It's possible to avoid the sleeping by using a condition variable to flag when the queue is empty

Reply:
I added a note to investigate this.